File: Others/medium/CycleInGraph.py
def cycleInGraph(edges):

    in_deg = [0] * len(edges)

    for vertex in edges:
        for adj in edges[vertex]:
            in_deg[adj] += 1

    queue = [x for x in range(len(edges)) if in_deg[x] == 0]

    while queue:
        current_node = queue.pop(0)
        for adj in edges[current_node]:
            in_deg[adj] -= 1
            if in_deg[adj] == 0:
                queue.append(adj)

    return sum(in_deg) == 0

File: Golang/medium/cycle-in-graph.go
package main

func DFS(graph [][]int, node int, visited []bool, recStack []bool, parent int) bool {
	visited[node] = true
	recStack[node] = true

	for _, adj := range graph[node] {
		if !visited[adj] {
			if DFS(graph, adj, visited, recStack, node) {
				return true
			}
		} else if recStack[adj] {
			return true
		}
	}

	recStack[node] = false
	return false
}

func CycleInGraph(edges [][]int) bool {
	visited := []bool{}
	recStack := []bool{}
	for idx := 0; idx < len(edges); idx++ {
		visited = append(visited, false)
		recStack = append(recStack, false)
	}

	for node := 0; node < len(edges); node++ {
		if !visited[node] {
			if DFS(edges, node, visited, recStack, -1) {
				return true
			}
		}
	}

	return false
}
